virttest/utils_test/__init__.py

```diff
@@ -139,49 +168,64 @@ def update_boot_option(vm, args_removed=None, args_added=None,
+                                                     remove_args=args_removed,
+                                                     args=args_added)
     if "ubuntu" in vm.get_distro().lower():
-        if args_added:
-            update_boot_option_ubuntu(args_added, session=session)
-        if args_removed:
-            update_boot_option_ubuntu(args_removed, session=session,
+        if req_args:
+            update_boot_option_ubuntu(req_args, session=session)
+        if req_remove_args:
+            update_boot_option_ubuntu(req_remove_args, session=session,
                                       remove_args=True)
     else:
         if not utils_package.package_install("grubby", session=session):
             raise exceptions.TestError("Failed to install grubby package")
         msg = "Update guest kernel option. "
         cmd = "grubby --update-kernel=`grubby --default-kernel` "
-        if bool(args_removed):
-            msg += " remove args: %s" % args_removed
-            cmd += '--remove-args="%s" ' % args_removed
-        if bool(args_added):
-            msg += " add args: %s" % args_added
-            cmd += '--args="%s"' % args_added
-        error_context.context(msg, logging.info)
-        status, output = session.cmd_status_output(cmd)
-        if status != 0:
-            logging.error(output)
-            raise exceptions.TestError("Fail to modify guest kernel option")
-
-    if need_reboot:
+        if req_remove_args:
+            msg += " remove args: %s" % req_remove_args
+            cmd += '--remove-args="%s" ' % req_remove_args
+        if req_args:
+            msg += " add args: %s" % req_args
+            cmd += '--args="%s"' % req_args
+        if req_remove_args or req_args:
+            error_context.context(msg, logging.info)
+            status, output = session.cmd_status_output(cmd)
+            if status != 0:
+                logging.error(output)
+                raise exceptions.TestError("Fail to modify guest kernel option")
+
+    # reboot is required only if we really add/remove any args
+    if need_reboot and (req_args or req_remove_args):
         error_context.context("Rebooting guest ...", logging.info)
         session = vm.reboot(session=session, timeout=login_timeout)
-    cmdline = session.cmd("cat /proc/cmdline", timeout=60)
-    if args_removed and args_removed in cmdline:
-        err = "Fail to remove guest kernel option %s" % args_removed
-        raise exceptions.TestError(err)
-    if args_added and args_added not in cmdline:
-        err = "Fail to add guest kernel option %s" % args_added
-        raise exceptions.TestError(err)
+        # check nothing is required to be added/removed by now
+        req_args, req_remove_args = check_kernel_cmdline(session,
+                                                         remove_args=args_removed,
+                                                         args=args_added)
+        if session:
+            session.close()
```

Comment:
Not necessary since already had line 227 and 228.

Reply:
it would close the session before exception raised in L223 or L226.

Reply:
Well, `exceptions.TestError` can be raised at line 208 where without a cleanup of the session. So if you want the session always to be cleaned up, the code would be:
```python
session = vm.wait_for_login(timeout=login_timeout)
try:
    ...
finally:
    if session:
        session.close()
```

Reply:
you are right, I have adopted your suggestion.